src/components/FirebaseFunctionsPageTable.tsx

```diff
@@ -0,0 +1,205 @@
+        {row.status!}
+      </Typography>
+    ),
+  },
+  {
+    title: 'Region',
+    field: 'region',
+    width: '150px',
+    render: (row: Partial<FunctionData>) => (
+      <Typography variant="body2" noWrap>
+        {row.region!}
+      </Typography>
+    ),
+  },
+  {
+    title: 'Last modified',
+    field: 'updateTime',
+    width: '150px',
+    render: (row: Partial<FunctionData>) => (
+      <Typography variant="body2" noWrap>
+        {getElapsedTime(row.updateTime!)}
+      </Typography>
+    ),
+  },
+  {
+    title: 'Runtime',
+    field: 'runtime',
+    width: '150px',
+    render: (row: Partial<FunctionData>) => (
+      <Typography variant="body2" noWrap>
+        {row.runtime}
+      </Typography>
+    ),
+  },
+  {
+    title: 'Memory',
+    field: 'availableMemoryMb',
+    width: '150px',
+    render: (row: Partial<FunctionData>) => (
+      <Typography variant="body2" noWrap>
+        {row.availableMemoryMb} MB
+      </Typography>
+    ),
+  },
+  {
+    title: 'Logs',
+    field: '',
+    width: '150px',
+    render: (row: Partial<FunctionData>) => {
+      return (
+        <Link
+          href={`https://console.cloud.google.com/logs/viewer?project=${row.project}&resource=cloud_function%2Ffunction_name%2F${row.name}%2Fregion%2F${row.region}`}
+          target="_blank"
+        >
+          view logs
+        </Link>
+      );
+    },
+  },
+];
+
+export const FirebaseFunctionsPageTable: React.FC = () => {
+  const [page, setPage] = useState(0);
+  const [pageSize, setPageSize] = useState(5);
+  const [filteredRows, setFilteredRows] = useState<FunctionData[]>([]);
+  const [settings] = useSettings();
+  const [tableProps] = useFirebaseFunctions({
+    project: settings.project,
+    authMethod: settings.authMethod,
+    apiKey: settings.apiKey,
+  });
+
+  useEffect(() => {
+    tableProps.retry();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [settings.authMethod, settings.project]);
+
+  useEffect(() => {
+    setFilteredRows(
+      tableProps.functionsData?.slice(page * pageSize, (page + 1) * pageSize) ??
+        [],
+    );
+  }, [tableProps.functionsData, page, pageSize]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  return (
+    <Table
+      isLoading={tableProps.loading || tableProps.loading}
+      options={{
+        paging: true,
+        pageSize,
+
+        padding: 'dense',
+        paginationType: 'normal',
+      }}
+      totalCount={tableProps.functionsData?.length ?? 0}
+      page={page}
+      data={filteredRows ?? []}
+      onChangePage={setPage}
+      onChangeRowsPerPage={setPageSize}
+      title={
+        <>
+          <Box display="flex" alignItems="center">
+            <Box mr={1} />
+            <Typography variant="h6">{settings.project}</Typography>
+          </Box>
+        </>
+      }
+      columns={columnDefinitions}
+      detailPanel={(rowData: FunctionData) => {
+        return (
+          <Box display="flex" p={1}>
```

Comment:
I think it's worth splitting this out to a separate component for readability.

Reply:
Done